File: raspberry/SmartHome/app/src/main/java/ru/smarthome/server/http/handlers/ControllerGet.java
package ru.smarthome.server.http.handlers;

import android.support.annotation.NonNull;

import com.google.gson.Gson;

import java.io.IOException;
import java.util.Map;

import fi.iki.elonen.NanoHTTPD;
import ru.smarthome.arduinodevices.ArduinoControllerResponse;
import ru.smarthome.arduinodevices.controllers.ArduinoReadable;
import ru.smarthome.library.BaseController;

public class ControllerGet extends BaseRequestHandler{

    @Override
    public NanoHTTPD.Response serve(NanoHTTPD.IHTTPSession session) {
        Map<String, String> params = session.getParms();
        BaseController controller = getController(params);
        if (controller instanceof ArduinoReadable) {
            try {
                ArduinoControllerResponse response = ((ArduinoReadable) controller).read();
                return new NanoHTTPD.Response(NanoHTTPD.Response.Status.OK, "text/json",
                        new Gson().toJson(response));
            } catch (IOException e) {
                return getArduinoHttpError();
            }
        } else {
            throw new IllegalStateException("get request to non readable service");
        }
    }
}

File: raspberry/SmartHome/arduinodevices/src/main/java/smarthome/raspberry/arduinodevices/server/httphandlers/ControllerGet.kt
package smarthome.raspberry.arduinodevices.server.httphandlers

import com.google.gson.Gson
import fi.iki.elonen.NanoHTTPD
import smarthome.raspberry.arduinodevices.ArduinoDeviceChannel

internal class ControllerGet : BaseRequestHandler() {
    private val channel: ArduinoDeviceChannel = TODO()

    override suspend fun serve(session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response {
        val params = session.parms
        val controller = getController(params)
        val device = input.findDevice(controller)

        val state = channel.read(device, controller)
        return NanoHTTPD.Response(NanoHTTPD.Response.Status.OK, "text/json",
                Gson().toJson(state))
    }
}
